File: src/src_platform_android/fmg/android/img/MosaicSkillImg.java
package fmg.android.img;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import fmg.common.Color;
import fmg.common.Pair;
import fmg.common.geom.PointDouble;
import fmg.core.img.IImageController;
import fmg.core.img.MosaicSkillController;
import fmg.core.img.MosaicSkillModel;
import fmg.core.types.ESkillLevel;

/**
 * Representable {@link ESkillLevel} as image
 * <br>
 * Android impl
 *
 * @param <TImage> Android specific image: {@link android.graphics.Bitmap})
 **/
public abstract class MosaicSkillImg<TImage> extends MosaicSkillOrGroupView<TImage, MosaicSkillModel> {

    /** @param skill - may be null. if Null - representable image of ESkillLevel.class */
    protected MosaicSkillImg(ESkillLevel skill) {
        super(new MosaicSkillModel(skill));
    }

    @Override
    protected Stream<Pair<Color, Stream<PointDouble>>> getCoords() { return getModel().getCoords(); }

    @Override
    public void close() {
        getModel().close();
        super.close();
    }

    /////////////////////////////////////////////////////////////////////////////////////////////////////
    //    custom implementations
    /////////////////////////////////////////////////////////////////////////////////////////////////////

    /** MosaicsSkill image view implementation over {@link android.graphics.Bitmap} */
    static class Bitmap extends MosaicSkillImg<android.graphics.Bitmap> {

        private BmpCanvas wrap = new BmpCanvas();

        public Bitmap(ESkillLevel skill) { super(skill); }

        @Override
        protected android.graphics.Bitmap createImage() {
            return wrap.createImage(getModel().getSize());
        }

        @Override
        protected void drawBody() {
            draw(wrap.getCanvas());
        }

        @Override
        public void close() {
            wrap.close();
        }

    }

    /** MosaicsSkill image controller implementation for {@link Bitmap} */
    public static class ControllerBitmap extends MosaicSkillController<android.graphics.Bitmap, Bitmap> {

        public ControllerBitmap(ESkillLevel skill) {
            super(skill == null, new MosaicSkillImg.Bitmap(skill));
        }

        @Override
        public void close() {
            getView().close();
            super.close();
        }

    }

}

File: src/src_platform_android/fmg/android/img/MosaicSkillImg.kt
package fmg.android.img

import java.util.stream.Stream

import fmg.common.Color
import fmg.common.Pair
import fmg.common.geom.PointDouble
import fmg.core.img.MosaicSkillController
import fmg.core.img.MosaicSkillModel
import fmg.core.types.ESkillLevel

/**
 * Representable [ESkillLevel] as image
 * <br></br>
 * Android impl
 *
 * @param TImage Android specific image: [android.graphics.Bitmap]
 */
internal abstract class MosaicSkillImg<TImage>
    /** @param skill - may be null. if Null - representable image of ESkillLevel.class */
    protected constructor(skill: ESkillLevel?) : MosaicSkillOrGroupView<TImage, MosaicSkillModel>(MosaicSkillModel(skill))
{

    override val coords: Stream<Pair<Color, Stream<PointDouble>>>
        get() = model.coords

    override fun close() {
        model.close()
        super.close()
    }

    /////////////////////////////////////////////////////////////////////////////////////////////////////
    //    custom implementations
    /////////////////////////////////////////////////////////////////////////////////////////////////////

    /** MosaicsSkill image view implementation over [android.graphics.Bitmap]  */
    class Bitmap(skill: ESkillLevel?) : MosaicSkillImg<android.graphics.Bitmap>(skill) {

        private val wrap = BmpCanvas()

        override fun createImage(): android.graphics.Bitmap {
            return wrap.createImage(model.size)
        }

        override fun drawBody() {
            draw(wrap.canvas)
        }

        override fun close() {
            wrap.close()
        }

    }

    /** MosaicsSkill image controller implementation for [Bitmap]  */
    class ControllerBitmap(skill: ESkillLevel?) : MosaicSkillController<android.graphics.Bitmap, Bitmap>(skill == null, MosaicSkillImg.Bitmap(skill)) {

        override fun close() {
            view.close()
            super.close()
        }

    }

}
